core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -376,7 +376,7 @@ public boolean isPreloaded() {
       final AdvancedCache<Object, Object> flaggedCache = getCacheForStateInsertion();
       return Flowable.fromPublisher(preloadCl.entryPublisher(null, true, true))
             .take(maxEntries)
-            .observeOn(nonBlockingScheduler)
+            .observeOn(blockingScheduler)
             .doOnNext(me -> preloadKey(flaggedCache, me))
```

Comment:
```suggestion
            .doOnNext(me -> preloadKey(flaggedCache, me))
            .observeOn(nonBlockingScheduler)
```

We already subscribe on the `blockingScheduler` below so there is no reason to invoke `observeOn` with the same value. We should really observe the completion on a non blocking thread since it is highly likely we may block on the completion in a blocking thread at some point.

And technically it should do observeOn before hand if it isn't tx, but we can avoid that for now.


While this change is fine with me, this really isn't a good idea to continue this on the blocking executor. This should really be done in a non blocking way if possible (ie. putAsync). I have logged https://issues.redhat.com/browse/ISPN-11387 to handle this better later when we can.

Reply:
done!

Reply:
Actually, sorry Pedro, but can you move it down 1 more line. So the `observeOn` is after the count. Otherwise we will have up to N-1  context switches just to add a counter. After the counter it will only send 1 value downstream.

Reply:
updated! :+1: